Add a script to sequentially scan and resync corrupted resources

Millefeuille42 opened the pull request:
Scans sequentially using drbdadm verify on each host and elects a "corrupted" host based on the reports. The resource is then resync-ed with drbdadm invalidate-remote.

Millefeuille42 committed:
Add a script to sequentially scan and resync corrupted resources

Scans sequentially using drbdadm verify on each host and elects
a "corrupted" host based on the reports. The resource is then
resync-ed with drbdadm invalidate-remote.

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

klmp200 reviewed (commented):
I don't see an issue with the logic
I only have minor cosmetic complains
That's good work :)

Millefeuille42 commented on sources/drbd_resource_repair.py:
```diff
@@ -0,0 +1,296 @@
+    )
+
+
+def verify_resource(status: ResourceStatus, remote_host: str = "") -> None:
```
By itself the `drbdadm verify` command doesn't return anything, only way is to request `drbdsetup status` or to parse the kernel logs, but I don't want to query drbd too often (which is why I am not requesting status of each resource individually).

Additionally since this is a helper function wrapping the call, I want the name to be as close as possible to the call's name or use.

Millefeuille42 commented on sources/drbd_resource_repair.py:
```diff
@@ -0,0 +1,296 @@
+        remote_host = host.ip if host.ip != host.from_host.ip else ""
+        statuses = get_oos_status(lazy=True, resource=resource, remote_host=remote_host)
+
+        for s in statuses:
```
This should not be an issue with single line loops / inlined loops where context of the variable is easily inferable from the same line / line above IMHO. I won't fight over it though, even more if the general opinion prefers full variable names.

Millefeuille42 replied:
Changed in [84871e8](https://github.com/xcp-ng/xcp-ng-linstor/pull/2/commits/84871e8dc1649ccf0c29086a73cb9c1a739150fe)

Millefeuille42 committed:
fix: apply suggestions from reviews

- Rename one letter variables
- Rename status_list to statuses

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix: use script name for logging

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 commented on sources/drbd_resource_repair.py:
```diff
@@ -0,0 +1,294 @@
+
+from typing import Any, Dict, Iterator, List, Optional, Set, Union
+
+SCRIPT_NAME = "drbd_resource_repair"
```
fixed in [be502dd](https://github.com/xcp-ng/xcp-ng-linstor/pull/2/commits/be502dd8333ba22d9e8195ace1f991530dba0926)

Millefeuille42 committed:
fix: change drbd command wrappers names to be more explicit

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix: add remote_host on the DRBD_WAIT_SYNC call for resync

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Wescoeur commented on sources/drbd_resource_repair.py:
```diff
@@ -0,0 +1,292 @@
+)
+
+
+def run_command(cmd: str, ignore_dry_run: bool = False, remote_host: str = "") -> str:
```
Perhaps `dry_run` instead to simplify reading.

Wescoeur commented on sources/drbd_resource_repair.py:
```diff
@@ -0,0 +1,292 @@
+            isinstance(other, ResourceStatus)
+            and self.resource == other.resource
+            and self.peer == other.peer
+            and self.out_of_sync == other.out_of_sync
```
I think you can remove the last comparison, the key of a resource status should be a resource name and a peer name. In fact you can even have a `ResourceStatusKey` or something like that to split the key and the associated data.

Millefeuille42 committed:
fix: apply suggestions from review

- add _CMD_ to command constants
- add run_write/read_command to abstract ignore_dry_run
- add FIXME about changing from lru_cache to cache
- move logger declaration at the top of the file
- move name and peer from ResourceStatus to a parent class
- remove IP from Host eq and hash
- remove lazy param
- remove out_of_sync from ResourceStatus eq and hash
- rename resource to resource_name / name
- rename status to resource
- rename get_all_hosts to get_all_peers
- change `type: ignore` comment to a cast
- inline os.getpid call for logging.basicConfig

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
misc: rename drbd_resource_repair.py to drbd-resync

- add shebang
- make script executable

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
doc: add documentation on complex parts

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix: Remove StatusUnion and apply JSON type where it fits

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
feat: add exclude list and change resource to include list

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
refactor(drbd-resync): move command execution to a singleton CommandRunner

- add master checking on each command run
- change union to optional typing
- freeze dict instead of using __dict__ in to_json()
- better error handling
- minor style fixes

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
refactor: use Counter instead of dict to count resources

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
misc: use lazy formatting in logger

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
feat: add options for the script to be used as a service

- add a ssh-key flag to specify the ssh key to be used
- add a service flag to change how the logs are formatted
- add an output flag to specify the output file of the report

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix: remove xapi dependency

Do not ask xapi for the master and instead check the pool.conf file

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>
